src/main/java/co/rsk/federate/FederationProviderFromFederatorSupport.java

```diff
@@ -91,13 +98,12 @@ public Address getActiveFederationAddress() {
     @Override
     public Optional<Federation> getRetiringFederation() {
         Integer federationSize = federatorSupport.getRetiringFederationSize();
-        Optional<Address> optionalRetiringFederationAddress = getRetiringFederationAddress();
 
-        if (federationSize == -1 || !optionalRetiringFederationAddress.isPresent()) {
+        if (federationSize == FEDERATION_NON_EXISTENT.getCode()) {
             return Optional.empty();
         }
 
-        Address retiringFederationAddress = optionalRetiringFederationAddress.get();
+        Address retiringFederationAddress = getRetiringFederationAddress().orElse(null);
```

Comment:
That is true, I have updated it to throw an IllegalStateException and added a unit test for it. Let me know if this looks good.

Reply:
Looking good but not sure about catching the exception and returning empty. It is an unexpected state, should we maybe let the exception propagate so we are informed that something is wrong?

If the retiring federation exists (members > -1) and we can't get the address something is very wrong. Returning empty implies there is no retiring federation and I don't think that's the right behavior

Reply:
Yes we can do this and I checked that this should be fine to let it propagate since the listener is wrapped in a try-catch so it won't bring the node down:

```java
    private void scheduleListenerCallbacks(Consumer<EthereumListener> callback) {
             for (EthereumListener listener : listeners) {
                 try {
                    callback.accept(listener);
                } catch (Throwable e) {
                     logger.error("Listener callback failed with exception", e);
                    panicProcessor.panic("thread", String.format("Listener callback failed with exception %s", e.getMessage()));
                }
            }
        }
```
Just notice that it will keep failling per onBestBlock which might be a good thing.

I have also removed it then for the Proposed federation. And updated the tests involved with this.